leakpro/attacks/mia_attacks/yoqo.py

```diff
@@ -0,0 +1,311 @@
+                )
+            )
+            self.in_indices_masks = np.concatenate(
+                (
+                    self.in_indices_masks[self.in_members,:],
+                    self.in_indices_masks[self.out_members,:]
+                ),
+                axis = 0
+            )
+
+    def _prepare_optimization_objective(self:Self) -> None:
+        if self.online:
+            self.weights = 1 * self.in_indices_masks + self.alpha * np.logical_not(self.in_indices_masks)
+        else:
+            self.weights = np.ones(self.in_indices_masks.shape)
+            self.mse = MSELoss()
+        self.weights = Tensor(self.weights)
+        self.batch_size = 32
+
+        data_loader = self.handler.get_dataloader(self.audit_data_indices, batch_size = self.batch_size)
+        self.target_output = []
+        for i, (data, labels) in tqdm(enumerate(data_loader),
+                                      total = len(data_loader),
+                                      desc="Calculating target labels",
+                                      leave=False):
+            if self.binary_output:
+                # Associate true labels with in-models and false labels with out-models.
+                batch_target_output = \
+                    labels[:,np.newaxis] * self.in_indices_masks[(i * self.batch_size):((i + 1) * self.batch_size),:] + \
+                    (1 - labels[:,np.newaxis]) * (1 - self.in_indices_masks[(i * self.batch_size):((i + 1) * self.batch_size),:])
+                self.target_output.extend(batch_target_output)
+            else:
+                # Associate true labels with in-models, and
+                # erroneous labels with largest logit score to out-models.
+                labels_true = labels
+                labels_false = torch.stack([shadow_model.model_obj(data) for shadow_model in self.shadow_models])
+                labels_false[:,np.arange(labels_true.shape[0]),labels_true] = -1e10
+                labels_false = torch.argmax(labels_false, dim = 2)
+                batch_target_output = \
+                    labels_true[:,np.newaxis] * self.in_indices_masks[(i * self.batch_size):((i + 1) * self.batch_size),:] + \
+                    labels_false.T * (1 - self.in_indices_masks[(i * self.batch_size):((i + 1) * self.batch_size),:])
+                self.target_output.extend(batch_target_output)
+        self.target_output = np.array(self.target_output)
+        self.target_output = Tensor(self.target_output)
+
+    def _optimization_objective(
+        self: Self,
+        x0: Tensor,
+        dx: Tensor,
+        target_output: Tensor,
+        weights: Tensor
+    ) -> float:
+        if self.binary_output:
+            loss = torch.cat([shadow_model.model_obj(x0 + dx) for shadow_model in self.shadow_models], dim = 1)
+        else:
+            loss = torch.stack([shadow_model.model_obj(x0 + dx) for shadow_model in self.shadow_models], dim = 2)
+        loss = self.loss(loss, target_output)
```

Comment:
For Binary cross entropy, this crashes as target_output must be float.
For Crossentropy, the target should be long so it works.